mitm/mitm.go

```diff
@@ -0,0 +1,281 @@
+// Copyright 2015 Google Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package mitm provides tooling for MITMing TLS connections. It provides
+// tooling to create CA certs and generate TLS configs that can be used to MITM
+// a TLS connection with a provided CA certificate.
+package mitm
+
+import (
+	"bytes"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/sha1"
+	"crypto/tls"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"errors"
+	"log"
+	"math/big"
+	"net"
+	"sync"
+	"time"
+)
+
+// MaxSerialNumber is the upper boundary that is used to create unique serial
+// numbers for the certificate. This can be any unsigned integer up to 20
+// bytes (2^(8*20)-1).
+var MaxSerialNumber = big.NewInt(0).SetBytes(bytes.Repeat([]byte{255}, 20))
+
+type Config struct {
+	// ca is the CA certificate used to sign MITM certificates.
+	ca *x509.Certificate
+	// capriv is the private key of the CA used to sign MITM certificates.
+	capriv interface{}
+	// priv is the private key used to generate certificates on the fly.
+	priv *rsa.PrivateKey
+	// keyID is the PKIX Subject Key ID for generated certificates.
+	keyID []byte
+	// validity is the window of time around time.Now() that the
+	// certificate will be valid.
+	validity time.Duration
+	// org that is displayed as the owner of the certificate.
+	org string
+	//
+	getCertificate func(*tls.ClientHelloInfo) (*tls.Certificate, error)
+	roots          *x509.CertPool
+
+	certmu sync.RWMutex
+	certs  map[string]*tls.Certificate
+}
+
+// NewAuthority creates a new CA certificate and associated
+// private key.
+func NewAuthority(name, organization string, validity time.Duration) (*x509.Certificate, *rsa.PrivateKey, error) {
+	priv, err := rsa.GenerateKey(rand.Reader, 2048)
+	if err != nil {
+		return nil, nil, err
+	}
+	pub := priv.Public()
+
+	// Subject Key Identifier support for end entity certificate.
+	// https://www.ietf.org/rfc/rfc3280.txt (section 4.2.1.2)
+	pkixpub, err := x509.MarshalPKIXPublicKey(pub)
+	if err != nil {
+		return nil, nil, err
+	}
+	h := sha1.New()
+	h.Write(pkixpub)
+	keyID := h.Sum(nil)
+
+	serial, err := rand.Int(rand.Reader, MaxSerialNumber)
```

Comment:
add a comment explaining that you know this isn't how you're supposed to create serial numbers - this is not a long lived cert, so it's "fine".